src/descriptor/mod.rs

```diff
@@ -66,6 +66,170 @@ pub use self::key::{
 /// public key from the descriptor.
 pub type KeyMap = HashMap<DescriptorPublicKey, DescriptorSecretKey>;
 
+/// A general trait for Bitcoin descriptor.
+/// It should also support FromStr, fmt::Display and should be liftable
+/// to bitcoin Semantic Policy.
+/// Offers function for witness cost estimation, script pubkey creation
+/// satisfaction using the [Satisfier] trait.
+// Unfortunately, the translation function cannot be added to trait
+// because of traits cannot know underlying generic of Self.
+// Thus, we must implement additional trait for translate function
+pub trait DescriptorTrait<Pk: MiniscriptKey>:
+    FromStr + Display + Debug + Clone + Eq + PartialEq + PartialOrd + Ord + Liftable<Pk>
+{
+    /// Whether the descriptor is safe
+    /// Checks whether all the spend paths in the descriptor are possible
+    /// on the bitcoin network under the current standardness and consensus rules
+    /// Also checks whether the descriptor requires signauture on all spend paths
+    /// And whether the script is malleable.
+    /// In general, all the guarantees of miniscript hold only for safe scripts.
+    /// All the analysis guarantees of miniscript only hold safe scripts.
+    /// The signer may not be able to find satisfactions even if one exists
+    fn sanity_check(&self) -> Result<(), Error>;
+
+    /// Computes the Bitcoin address of the descriptor, if one exists
+    /// `to_pk_ctx` denotes the ToPkCtx required for deriving bitcoin::PublicKey
+    /// from MiniscriptKey using [ToPublicKey].
+    /// If MiniscriptKey is already is [bitcoin::PublicKey], then the context
+    /// would be [NullCtx] and [descriptor.DescriptorPublicKeyCtx] if MiniscriptKey is [descriptor.DescriptorPublicKey]
+    ///
+    /// In general, this is defined by generic for the trait [trait.ToPublicKey]
+    fn address<ToPkCtx: Copy>(
+        &self,
+        to_pk_ctx: ToPkCtx,
+        network: bitcoin::Network,
+    ) -> Option<bitcoin::Address>
```

Comment:
When would we ever return None ? Especially since below we don't return an Option for the scriptPubKey.

Reply:
In the case of Bare Descriptor, we return `None`. But as another PR, after this is merged I plan to change all option returning APIs to return Results.